// playright/dist/index.js
"use strict";
// MIT License
//
// Copyright (c) 2020 Iakiv Kramarenko, Alexander Popov
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stage = exports.Elements = exports.Element = exports.be = exports.have = exports.perform = void 0;
var commands_1 = require("./commands");
Object.defineProperty(exports, "perform", { enumerable: true, get: function () { return commands_1.perform; } });
var conditions_1 = require("./conditions");
Object.defineProperty(exports, "have", { enumerable: true, get: function () { return conditions_1.have; } });
Object.defineProperty(exports, "be", { enumerable: true, get: function () { return conditions_1.be; } });
var element_1 = require("./element");
Object.defineProperty(exports, "Element", { enumerable: true, get: function () { return element_1.Element; } });
var elements_1 = require("./elements");
Object.defineProperty(exports, "Elements", { enumerable: true, get: function () { return elements_1.Elements; } });
var stage_1 = require("./stage");
Object.defineProperty(exports, "Stage", { enumerable: true, get: function () { return stage_1.Stage; } });
